src/main/java/net/fabricmc/loader/impl/discovery/ModDiscoverer.java

```diff
@@ -16,374 +16,471 @@
-				// normalize URL (used as key for nested JAR lookup)
-				normalizedUrl = UrlUtil.asUrl(path);
-			} catch (UrlConversionException e) {
-				throw new RuntimeException("Failed to convert URL " + url + "!", e);
-			}
-
-			if (Files.isDirectory(path)) {
-				// Directory
-				modJson = path.resolve("fabric.mod.json");
-				rootDir = path;
-
-				if (loader.isDevelopmentEnvironment() && !Files.exists(modJson)) {
-					Log.warn(LogCategory.DISCOVERY, "Adding directory %s to mod classpath in development environment - workaround for Gradle splitting mods into two directories", path);
-					synchronized (launcherSyncObject) {
-						FabricLauncherBase.getLauncher().propose(url);
-					}
-				}
-			} else {
-				// JAR file
-				try {
-					jarFs = FileSystemUtil.getJarFileSystem(path, false);
-					modJson = jarFs.get().getPath("fabric.mod.json");
-					rootDir = jarFs.get().getRootDirectories().iterator().next();
-				} catch (IOException e) {
-					throw new RuntimeException("Failed to open mod JAR at " + path + "!");
-				} catch (ZipError e) {
-					throw new RuntimeException("Jar at " + path + " is corrupted, please redownload it!");
+				if (is != null) {
+					return computeJarStream();
+				} else if (Files.isDirectory(path)) {
+					return computeDir();
+				} else {
+					return computeJarFile();
 				}
+			} catch (IOException | ParseMetadataException e) {
+				throw ExceptionUtil.wrap(e);
 			}
+		}
 
-			LoaderModMetadata[] info;
+		private ModCandidate computeDir() throws IOException, ParseMetadataException {
+			Path modJson = path.resolve("fabric.mod.json");
+			if (!Files.exists(modJson)) return null;
 
-			try {
-				info = new LoaderModMetadata[] { ModMetadataParser.parseMetadata(modJson) };
-			} catch (ParseMetadataException.MissingRequired e) {
-				throw new RuntimeException(String.format("Mod at \"%s\" has an invalid fabric.mod.json file! The mod is missing the following required field!", path), e);
-			} catch (MalformedJsonException | ParseMetadataException e) {
-				throw new RuntimeException(String.format("Mod at \"%s\" has an invalid fabric.mod.json file!", path), e);
-			} catch (NoSuchFileException e) {
-				Log.warn(LogCategory.DISCOVERY, "Non-Fabric mod JAR at \"%s\", ignoring", path);
-				info = new LoaderModMetadata[0];
-			} catch (IOException e) {
-				throw new RuntimeException(String.format("Failed to open fabric.mod.json for mod at \"%s\"!", path), e);
-			} catch (Throwable t) {
-				throw new RuntimeException(String.format("Failed to parse mod metadata for mod at \"%s\"", path), t);
+			LoaderModMetadata metadata;
+
+			try (InputStream is = Files.newInputStream(modJson)) {
+				metadata = ModMetadataParser.parseMetadata(is, localPath, parentPaths);
 			}
 
-			for (LoaderModMetadata i : info) {
-				ModCandidate candidate = new ModCandidate(i, normalizedUrl, depth, requiresRemap);
-				boolean added;
+			if (!metadata.loadsInEnvironment(envType)) {
+				return null;
```

Comment:
Looking at #375, should we add a warning here? Though if a mod is not loaded, it won't appear in loader's mod list, and should be quite clear in that case.

Reply:
The mod explicitly opted out of being loaded, doesn't look warning worthy to me?

Reply:
Hmm, now I am thinking about using the 5-tier dependency system to handle the environment thing: a mod can declare "breaks" to strictly crash when loading on one side and "conflicts" for less-strict warning.

In fact, I think we can add an extra minecraft client builtin mod; the whole dedicated server is shipped within the client, so it will just count as part of `minecraft` builtin mod. If a mod opts-in to be client-only, it can declare a hard dependency on client mod; otherwise, it can declare a hard breaks on client (though I see no point for a mod to do so)

Reply:
Hmm, @Chocohead what's the main problems with this?

Now, after some thinking, I feel such a split by using jij for client-specific code is actually a good approach, somewhat more feasible than adding `@Environment` annotations and having less compile-time guarantees

Reply:
One would just declare the environment in the mod json instead of messing with dependencies

Reply:
I don't think we should be encouraging/offering people to crash if their mod is on the wrong side. Right now we avoid loading it which is a much more user friendly approach to the problem. Forge in comparison expects mods to handle it themselves (unlike in the past with the old `clientOnlyMod`) but also doesn't offer a way to crash on the wrong side.